src/Semaphore.sol

```diff
@@ -1,208 +1,377 @@
+import {Verifier as MerkleTreeVerifier} from "./generated/TreeVerifier.sol";
+
+/// @title WorldID Identity Manager
+/// @author Worldcoin
+/// @notice An implementation of a batch-based identity manager for the WorldID protocol.
+/// @dev The manager is based on the principle of verifying externally-created Zero Knowledge Proofs
+///      to perform the insertions.
+contract Semaphore is IWorldID, SemaphoreCore {
+    ///////////////////////////////////////////////////////////////////////////////
+    ///                       PUBLIC CONFIGURATION STORAGE                      ///
+    ///////////////////////////////////////////////////////////////////////////////
+
+    /// @notice The address that manages this contract, which is allowed to update and create groups.
+    address public manager = msg.sender;
 
-/// @title Semaphore Group Manager
-/// @author Miguel Piedrafita
-/// @notice A simple implementation of a ZK-based identity group manager using Semaphore
-contract Semaphore is IWorldID, SemaphoreCore, Verifier, SemaphoreGroups {
-    using IncrementalBinaryTree for IncrementalTreeData;
+    /// @notice The latest root of the merkle tree.
+    uint256 public latestRoot;
 
     ///////////////////////////////////////////////////////////////////////////////
-    ///                                  ERRORS                                 ///
-    //////////////////////////////////////////////////////////////////////////////
+    ///                             EXTERNAL CONSTANTS                          ///
+    ///////////////////////////////////////////////////////////////////////////////
 
-    /// @notice Thrown when trying to update or create groups without being the manager
-    error Unauthorized();
+    /// @notice A constant representing a root that doesn't exist.
+    /// @dev Can be checked against when querying for root data.
+    function NO_SUCH_ROOT() public pure returns (RootInfo memory rootInfo) {
+        return RootInfo(0x0, 0x0, false);
+    }
 
-    /// @notice Thrown when trying to create a group with id 0, since this can later cause issues with root history verification.
-    error InvalidId();
+    ///////////////////////////////////////////////////////////////////////////////
+    ///                               PUBLIC TYPES                              ///
+    ///////////////////////////////////////////////////////////////////////////////
 
-    /// @notice Thrown when attempting to validate a root that doesn't belong to the specified group.
-    error InvalidRoot();
+    /// @notice The proof data for verifying the merkle tree proof.
+    /// @dev Corresponds to `MerkleTreeVerifier.Proof` but provides an external interface.
+    ///
+    /// @param ar Corresponds to `ar` in the proof or `A` in the internal proof.
+    /// @param bs Corresponds to `bs` in the proof or `B` in the internal proof.
+    /// @param krs Corresponds to `krs` in the proof or `C` in the internal proof.
+    struct MerkleTreeProof {
```

Comment:
Not sure all of this is necessary. This data is of no interest to our contract, so we might as well take it in as `uint256[8] calldata` and save some gas.